design/ceph/nvmeof-recoverer.md

```diff
@@ -0,0 +1,184 @@
+
+So, what we propose is a functionality that moves the OSD pod to another server node either automatically or manually when the server node fails because storage devices are still workable in NVMe-oF environment.
+
+### Goals
+
+ ![integration](nvmeof-recoverer.png)
+
+- **Enhanced Fault Tolerance**: NVMe-oF allows storage to be disaggregated from specific server nodes, which makes it possible to dynamically reassign resources in case of compute node failures. This capability provides enhanced fault tolerance, as storage can be attached to any operational node without requiring physical movement of devices.
+- **Minimized Downtime**: In traditional systems, node failures are manually recovered or reattached. The Rook operator retries pod-level recovery for transient failures, but if repeated restarts fail, it marks the OSD as out, allowing Ceph-level recovery (rebalancing). NVMe-oF recoverer minimizes this downtime by automatically reassigning storage devices to other nodes, ensuring data availability with minimal delay even in the event of unrecoverable compute node failures.
+
+- **Load balance**: As storage needs grow, Recoverer enables easy expansion of storage capacity. It introduces a scheduler that automatically connects fabric devices to suitable nodes using load-balancing and deploys OSDs. By utilizing Kubernetes Jobs, the scheduler dynamically connects based on the workload ensuring optimal performance
+
+### Non-Goals
+
+- **Support for Non-NVMe-oF Environments**: This design focuses solely on NVMe-oF environments.
+- **Modifications to the Core Ceph Codebase**: The proposal does not involve changing Ceph's core functionalities.
+
+## Proposal Details
+
+### Overview
+
+We propose a design, called NVMe-oF recoverer. NVMe-oF is an emerging storage network where the storage devices can be attached to any storage node using RDMA, TCP or Fiber channel. In this environment, the failure caused by a server node can be recoverable because the OSD can be re-spawned to another server node with the storage device attached to the failed server node through NVMe-oF.
+
+To address these challenges, we recommend a phased implementation consisting of four key steps:
+
+1. **Implement an Explicit Command for Manual OSD Relocation**
+2. **Automate OSD Relocation Upon Failure**
+3. **Modify CRUSH Map to Minimize Data Rebalancing**
+4. **Improve OSD Scheduling Based on Node Workloads**
+
+Each step can be implemented and contributed individually, allowing for incremental progress and testing.
+
+### Step 1: Implement an Explicit Command for Manual OSD Relocation
+
+#### Description
+
+Introduce explicit commands or scripts that allow administrators to manually deploy and relocate OSD pods using NVMe-oF devices. These tools will streamline the management of NVMe-oF devices and OSD pods in standard scenarios and during failures.
+
+#### Implementation Steps
+
+- **Develop a Script or Command-line Tool**:
```

Comment:
IMO the only thing Rook should do is determine the OSD topology for the CRUSH map. The other details of this implementation seem like they would be part of a CSI driver.